Add possibility to override node package at cluster creation/update (only for development purposes)

jdeamicis opened the pull request:
### Description of changes
* Add possibility to install a custom node package at cluster creation/update (only for development purposes).
* Encapsulate installation of custom node package into specific recipe reused at AMI build, cluster creation and head node update.

### Tests
* Added some kitchen tests for installation of custom node at install phase and at cluster creation phase.

### References
* https://github.com/aws/aws-parallelcluster-cookbook/pull/1480

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] ~Check if documentation is impacted by this change.~

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Add possibility to override node package at cluster creation and update

Rename node.rb to parallelcluster_node.rb

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Adapt node chefspec test to custom node recipe

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add kitchen test for custom node installation

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Move custom_parallelcluster_node.rb to computefleet cookbook

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add kitchen test for installing custom node at finalize stage

Move update of node package into sub-recipe of update.rb

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Address cookstyle offence

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add changelog entry

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on cookbooks/aws-parallelcluster-entrypoints/recipes/finalize.rb:
```diff
@@ -16,6 +16,11 @@
 
 fetch_config 'Fetch and load cluster configs'
 
+if !node['cluster']['custom_node_package'].nil? && !node['cluster']['custom_node_package'].empty?
```
Since this is used in different part of the code, can we encapsulate this check in a helper function `custom_node_package?` this would make the code more readable and the recipes more maintainable.

jdeamicis replied:
Addressed.

NSsirena commented on cookbooks/aws-parallelcluster-entrypoints/recipes/update.rb:
```diff
@@ -22,3 +22,18 @@
+    action :stop
+  end
+
+  include_recipe 'aws-parallelcluster-shared::parallelcluster_custom_node'
```
Why in is defined in the shared cookbook ?
Do we need it in other cookbook ? And if so, why ?

NSsirena replied:
Ok.. I saw you moved it in a following commit.

NSsirena commented on cookbooks/aws-parallelcluster-computefleet/test/controls/custom_parallelcluster_node_spec.rb:
```diff
@@ -0,0 +1,28 @@
+  file_cache_path = "/tmp/kitchen/cache"
+
+  title "custom aws-parallelcluster-node should have been installed in the node virtualenv"
+  only_if { !os_properties.redhat_ubi? }
```
Why not ?

jdeamicis replied:
Good question. I actually started working from [node_spec.rb](https://github.com/aws/aws-parallelcluster-cookbook/blob/b4ef8f5c0b586af0f1e306ce5ea81700a192321b/cookbooks/aws-parallelcluster-computefleet/test/controls/node_spec.rb), which currently has that if clause as well. I'm not sure why: I'll try to remove it and see if it breaks on docker or RHEL.

NSsirena commented on cookbooks/aws-parallelcluster-computefleet/recipes/install/node.rb:
```diff
@@ -35,27 +35,7 @@
-      echo "PATH is $PATH"
-      source #{virtualenv_path}/bin/activate
-      pip uninstall --yes aws-parallelcluster-node
-      if [[ "#{node['cluster']['custom_node_package']}" =~ ^s3:// ]]; then
-        custom_package_url=$(#{cookbook_virtualenv_path}/bin/aws s3 presign #{node['cluster']['custom_node_package']} --region #{node['cluster']['region']})
-      else
-        custom_package_url=#{node['cluster']['custom_node_package']}
-      fi
```
[Nitpick] This could be moved into an helper function and then we can store `node_package_url`  as part of the pcluster config. 
This will allow us also to test if we have a custom node or not.

jdeamicis replied:
That's inside a bash block for the time being. I agree that we should rewrite that recipe in a more chef-friendly way, but I would keep it out of scope for now (it can be part of our cookbook refactoring later on).

NSsirena commented on cookbooks/aws-parallelcluster-computefleet/test/controls/custom_parallelcluster_node_spec.rb:
```diff
@@ -0,0 +1,28 @@
+  title "custom aws-parallelcluster-node should have been installed in the node virtualenv"
+  only_if { !os_properties.redhat_ubi? }
+
+  # Unless we fix the version of aws-parallelcluster-node to be installed, we cannot really say from pip if a custom
+  # node package was installed. The only thing we can test is that the custom node recipe was triggered, which can
+  # be verified from some artifacts left in the kitchen cache folder.
```
[Nitpick] As said in a comment for the `install/node.rb` ..if we store the `node_package_url` as part of the config we should be able to check if we are using a custom node version.

jdeamicis replied:
Yep. I tend to agree, but I cannot spend much more time on this at the moment. We will do it as part of the cookbook refactoring.

jdeamicis committed:
Fix location of recipe after rebase

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Create helper function is_custom_node?

Use the new helper function to detect whether a custom node package is
requested and trigger the custom node installer and update recipes.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
